include/aparse/parser_builder.hpp

```diff
@@ -0,0 +1,49 @@
+// Copyright: 2015 Mohit Saini
+// Author: Mohit Saini (mohitsaini1196@gmail.com)
+
+#ifndef _APARSE_PARSER_BUILDER_HPP_
+#define _APARSE_PARSER_BUILDER_HPP_
+
+#include <tuple>
+
+#include <quick/type_traits/function_type.hpp>
+
+#include "aparse/common_headers.hpp"
+#include "aparse/error.hpp"
+#include "aparse/parser.hpp"
+#include "aparse/internal_parser_builder.hpp"
+
+namespace aparse {
+
+class ParserGrammar {
+ public:
+  struct Rule {
+    Rule (const string& rule): rule_string(rule) {}
+    template<typename T>
+    Rule& Action(const T& action) {
```

Comment:
Looking at the methods, it does seem like the level of debugging is very low here. I am not sure if that is a good practice to follow. I think the level of logs being added should be revisited here to ensure that the code quality is high from a debuggibility point of view.

Reply:
In AParse, logging is not injected in code to prevent the code pollution. Instead of injection, almost all the internal classes have DebugStream/DebugString method. Debugger can simply use `#include <quick/debug.hpp>` and do `cout << object` to get the highly readable object structure.
There are strict validation at major gateways to detect the buggy component. A debugger can take out the buggy component and offending input/output of that component, and do the postmortem of offending component in isolation without touching the rest of components.   
Debugger have to inject logging for debugging purpose in isolated component. (if they want).


The task to update per-commit-check to "Detect the logging statements and prevent the change to merge in master" is still pending.